ql/src/java/org/apache/hadoop/hive/ql/history/HiveScribeImpl.java

```diff
@@ -0,0 +1,334 @@
+package org.apache.hadoop.hive.ql.history;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.hadoop.hive.ql.QueryPlan;
+import org.apache.hadoop.hive.ql.QueryStats;
+import org.apache.hadoop.hive.ql.exec.Task;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hadoop.mapred.Counters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HiveScribeImpl implements HiveHistory {
+  private static final Logger LOG = LoggerFactory.getLogger("hive.ql.exec.HiveScribeImpl");
+
+  private Map<String, String> idToTableMap = null;
+
+  // Job Hash Map
+  private final HashMap<String, HiveHistory.QueryInfo> queryInfoMap = new HashMap<String, HiveHistory.QueryInfo>();
+
+  // Task Hash Map
+  private final HashMap<String, HiveHistory.TaskInfo> taskInfoMap = new HashMap<String, HiveHistory.TaskInfo>();
+
+  private static final String ROW_COUNT_PATTERN = "RECORDS_OUT_(\\d+)(_)*(\\S+)*";
+
+  private static final Pattern rowCountPattern = Pattern.compile(ROW_COUNT_PATTERN);
+
+  // Query Hash Map
+  private static HashMap<String, QueryStats> queryStatsMap = new HashMap<String, QueryStats>();
+
+  /**
+   * Construct HiveScribeImpl object and scribe query metric to log pipeline.
+   *
+   * @param sessionState
+   */
+  public HiveScribeImpl(SessionState sessionState) {
+    LOG.info("Instantiated an instance for HiveScribeImpl to scribe query logs. \n");
+  }
+
+  /**
+   * HistFileName is not used in this impl.
+   */
+  @Override
+  public String getHistFileName() {
+    return "HiveQueryCompeletionScribe";
+  }
+
+  /**
+   * Call TwitterScriber to scribe hive query statistics to log pipeline
+   *
+   * @param queryID
+   */
+  private void scribeQueryMetricEntry (String queryID) {
+    QueryStats stats = queryStatsMap.get(queryID);
+    if (stats == null) {
+      LOG.info("Stats is null. Nothing passed to log pipeline");
+      return;
+    }
+
+    QueryCompletedEventScriber hiveHistScriber = new QueryCompletedEventScriber();
+    hiveHistScriber.handle(stats);
+    LOG.info("Query stats passed to log pipeline");
+    queryStatsMap.remove(queryID);
```

Comment:
remove this query from `queryStatsMap` in the caller.

Reply:
done